### ISSUE TITLE
🔍 Advisor Code Review: Architecture, Code Quality & Engineering Practices

### ISSUE THREAD

**Issue by syifan** [0nhc/m2sim#1]
## 🔍 Advisor Code Review: M2Sim Project

**Review Date:** February 2, 2026  
**Commit Reviewed:** d8d1cd2 ("update")  
**Repository:** 0nhc/m2sim

---

### 📋 Executive Summary

M2Sim is a trajectory prediction and simulation framework built on top of the M2I (Motion-to-Interaction) model, designed for autonomous driving scenarios using the Waymo Open Dataset. The project integrates deep learning-based trajectory prediction with Model Predictive Control (MPC) for motion planning.

The codebase shows promising research work but has significant opportunities for improvement in code quality, documentation, and maintainability.

---

### ✅ Strengths

1. **Clear Research Focus** - The project successfully combines M2I prediction with MPC control in a unified simulation loop
2. **Waymo Dataset Integration** - Proper integration with the Waymo Open Dataset ecosystem
3. **Configuration Management** - YAML-based configuration system allows flexible experimentation
4. **Performance Optimization** - Use of Cython (`utils_cython.pyx`) for computational bottlenecks

---

#### 1. **[ACTION] Hardcoded Absolute Paths**
```python
# main.py
path_whole = '/DATA1/liyang/M2I_2/data_split_new/' + MANUAL_SCENEE_NAME + '/'
m2i_checkpoint_path = '/DATA2/lpf/baidu/additional_files_for_m2i/test0908_0/model_save/model.12.bin'
```
**Impact:** Code is not portable; will fail immediately for any other user.  
**Recommendation:** Use environment variables or configuration files for paths.

#### 2. **[ACTION] Bare Exception Handling**
Found 12+ instances of `except:` without specific exception types:
```python
try:
    res = get_inputs_for_M2I(df, sdc_id)
except:
    print('Break when preparing outputs.')
    break
```
**Impact:** Silently swallows all errors including KeyboardInterrupt, making debugging extremely difficult.  
**Recommendation:** Catch specific exceptions (`except ValueError as e:`) and log stack traces.

#### 3. **[ACTION] Missing `.gitignore`**
Compiled artifacts checked into repo:
- `src/__pycache__/*.pyc`
- `src/utils_cython.cpython-310-x86_64-linux-gnu.so`
- `src/build/temp.linux-x86_64-cpython-310/utils_cython.o`

**Impact:** Repository bloat, potential cross-platform issues.  
**Recommendation:** Add comprehensive `.gitignore` and remove build artifacts from git history.

---

#### 4. **[ACTION] Monolithic main.py (1,548 lines)**
The main script contains:
- Configuration constants
- Data loading
- Utility functions
- Core simulation loop
- Visualization code

**Impact:** Hard to maintain, test, or extend.  
**Recommendation:** Refactor into modules:
```
m2sim/
├── config.py          # Configuration management
├── data/              # Data loading and processing
├── simulation/        # Core simulation loop
├── visualization/     # Plotting utilities
└── main.py            # Entry point only
```

#### 5. **[INFO] Global Mutable State**
`src/globals.py` uses module-level mutable globals:
```python
interactive_single_traj_waymo_pred: Optional[WaymoPred] = None
waymo_autoreg_pred = WaymoAutoregPred()
temp = {}
```
**Impact:** Hard to reason about state, difficult to parallelize, testing challenges.  
**Note:** This may be intentional for research iteration speed.

#### 6. **[INFO] Inconsistent Naming**
- `MANUAL_SCENEE_NAME` (typo: SCENE not SCENEE)
- Mix of `snake_case` and `camelCase` in variable names
- Some Chinese comments mixed with English

---

#### 7. **[ACTION] Poor Documentation Coverage**
| File | Functions | Docstrings |
|------|-----------|------------|
| main.py | 57 | 18 |
| utils.py | 93 | 7 |
| m2i_script.py | 11 | 0 |
| format_transform_new.py | 7 | 0 |

Only ~10% of functions have docstrings.  
**Recommendation:** Add docstrings to at least public functions explaining inputs, outputs, and purpose.

#### 8. **[ACTION] No Type Hints**
Only 2 functions out of 265+ have return type annotations.  
**Recommendation:** Add type hints progressively, especially to public interfaces:
```python
def velocity_yaw_future(old_decoded_example: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
```

#### 9. **[ACTION] No Test Suite**
Zero test files found (`test_*.py` or `*_test.py`).  
**Impact:** No automated verification of functionality; regressions go undetected.  
**Recommendation:** Start with integration tests for the core simulation loop.

#### 10. **[INFO] TODOs in Code**
8 TODO comments found, some critical:
```python
# TODO: check M2I inputs before calling M2I.
# TODO: remove objects whose past or current contain invalid values
```
These should be tracked as issues.

---

#### 11. **[ACTION] Version Conflicts in requirements.txt**
```
torch==2.2.0
tensorflow==2.11.0
waymo-open-dataset-tf-2-11-0
```
TensorFlow 2.11 requires specific CUDA versions that may conflict with PyTorch 2.2.  
**Recommendation:** 
- Document tested CUDA/cuDNN versions explicitly
- Consider using a conda environment file
- Add version ranges instead of exact pins where safe

#### 12. **[INFO] README Installation Incomplete**
README only shows basic git clone and Cython build. Missing:
- Python version requirements
- GPU/CUDA setup instructions  
- Data download instructions
- Example usage beyond installation

---

### 📊 Metrics Summary

| Metric | Value | Assessment |
|--------|-------|------------|
| Total Lines of Code | 7,643 | Medium-sized project |
| Functions | 265+ | Significant surface area |
| Docstring Coverage | ~10% | Poor |
| Type Hint Coverage | <1% | Very Poor |
| Test Coverage | 0% | None |
| Open TODOs | 8 | Needs attention |
| Commits (total) | 6 | Very early stage |

---

### 🎯 Recommended Priorities

**Immediate (This Week):**
1. Add `.gitignore` and clean repo history
2. Replace hardcoded paths with config/env variables
3. Fix bare `except:` blocks

**Short-term (2 Weeks):**
4. Add basic integration tests
5. Refactor `main.py` into modules
6. Document core functions

**Medium-term (1 Month):**
7. Add type hints to public APIs
8. Create proper setup.py/pyproject.toml
9. Add CI/CD pipeline with linting

---

### 📝 Notes for Authors

The core algorithmic work appears sound, and the integration of M2I with MPC is a reasonable architectural choice for this problem domain. The main issues are engineering practices that will become increasingly painful as the project grows.

Consider adopting:
- **Black** for code formatting
- **mypy** for type checking
- **pytest** for testing
- **Pre-commit hooks** for quality gates

Happy to discuss any of these recommendations in more detail.

**Comment by syifan** [comment]
My apologies. My bot ran crazy and posted to the wrong repo. Please delete.
